docs/errors.md

```diff
@@ -0,0 +1,249 @@
+
+- #### `MongoMissingCredentialsError`
+
+  - Thrown when a user fails to provide authentication credentials before attempting to connect to the mongo server.
+
+- #### `MongoDependencyError`
+  - Thrown when a required module or dependency is not present.
+
+### `MongoRuntimeError`
+
+This class represents errors which occur when the driver encounters unexpected input or reaches an unexpected/invalid internal state.
+This class should **never** be directly instantiated.
+
+_MongoRuntimeError children (pt 1)_
+![(MongoRuntimeError hierarchy tree part 1)](charts/imgs/MongoRuntimeError_0.svg)
+
+_MongoRuntimeError children (pt 2)_
+![(MongoRuntimeError hierarchy tree part 2)](charts/imgs/MongoRuntimeError_1.svg)
+
+_MongoRuntimeError children (pt 3)_
+![(MongoRuntimeError hierarchy tree part 3)](charts/imgs/MongoRuntimeError_2.svg)
+
+Children of `MongoRuntimeError` include:
+
+- #### `MongoInternalDriverError`
+
+  - Thrown when a catastrophic failure occurs in the driver.
+
+- #### `MongoTransactionError`
+
+  - Thrown when the user makes a mistake in the usage of transactions (e.g.: attempting to commit a transaction with a readPreference other than primary).
+
+- #### `MongoClientNotConnectedError`
+
+  - Thrown when the user attempts to operate on the data from a client that has not been connected to a MongoDB server instance.
+
+- #### `MongoKerberosClientConnectionError`
+
+  - Thrown when the user attempts to authenticate via Kerberos, but fails to connect to the Kerberos client.
+
+- #### `MongoEncryptionError`
+
+  - Thrown when an autoencrypter is requested but unavailable.
+
+- #### `MongoCompressionError`
+
+  - Thrown when the driver fails to compress data before sending it to the server.
+
+- #### `MongoDecompressionError`
+
+  - Thrown when the driver fails to decompress data received from the server
+
+- #### `MongoExpiredSessionError`
+
+  - Thrown when the user attempts to operate on a session that has expired or has been closed.
+
+- #### `MongoIOError`
+
+  - Thrown when the driver fails to read or write from a file.
+
+- #### `MongoParseError`
+
+  - Thrown when the driver fails to correctly parse otherwise properly supplied input.
+
+- #### `MongoResourceClosedError`
+  - Thrown when there is an attempt to access a resource which has already been or will be closed/destroyed.
+    - Children of this error class include:
+      - **`MongoServerClosedError`**: Thrown when an attempt is made to operate on a closed server.
+      - **`MongoStreamClosedError`**: Thrown when an attempt is made to operate on a closed stream.
+      - **`MongoTopologyClosedError`**: Thrown when an attempt is made to operate on a dropped, or otherwise unavailable, database.
+- #### `MongoCursorError`
+
+  - Thrown when the user incorrectly uses a cursor object.
+    - Children of this error class include:
+      - **`MongoInvalidCursorOperationError`**: Thrown when the user calls a function or method that is not supported on the current cursor.
```

Comment:
Would this arise because of the server version not supporting some functionality?

Reply:
@nbbeeken The case we had in mind with this specifically is trying to sort a tailable cursor, so it may need some more specificity in its naming, but it have to do with the sever version.

Reply:
It looks like most errors in the Cursor are thrown during initialization (in methods with `assertUninitialized(this);` at the top). I'd propose we just give those a general name like `MongoCursorInitializationError`.

Reply:
@emadum: `assertUnintialized(this)` appears pretty often so I've made `MongoCursorInUseError` to cover the case where a user attempts to add options onto an already existing cursor that has failed `assertUnintialized(this)`. 

I've also renamed `MongoInvalidCursorOperationError` to `MongoTailableCursorError` because that was the only place where this error was being used. It will still be thrown when a user attempt to call a function that a TailableCursor doesn't support.